lib/csv_helper/loader.rb

```diff
@@ -19,9 +19,22 @@ def load(file, options = {})
     def load_records(file, options)
       records = []
 
-      records = klass == Institution ? load_csv_with_version(file, records, options) : load_csv(file, records, options)
+      records = if klass == Program
+                  load_csv_with_row(file, records, options)
+                elsif klass == Institution
+                  load_csv_with_version(file, records, options)
+                else
+                  load_csv(file, records, options)
+                end
+
       results = klass.import records, ignore: true, batch_size: Settings.active_record.batch_size.import
-      after_import_validations(records, results.failed_instances, options)
+
+      # using index of -1 since these rows failed during save to the table and not during after_import_validations
+      failed_instances = results.failed_instances
+                                .map { |result| { index: -1, message: result.display_errors_with_row } }
+      after_import_validations(records, failed_instances, options)
+
+      results.failed_instances = failed_instances
       results
```

Comment:
I don't like that we've now changed `failed_instances` from an array of AR objects (e.g. `Program` objects) to an array of hashes with warning data at this point.  I know we don't actually use those objects today but it might trip up someone in the future.
I'd rather either monkey patch `ActiveRecord::Import::Result` or just return our warnings array with the `results`

```ruby
return [results, import_warnings]
```

I also think you should rename the array from `failed_instances` to something like `import_warnings` since they are not actually failed instances.

Reply:
I have renamed the array that is being passed around for `after_import_batch_validations` to `validation_warnings`.

I have changed the object that is in `results.failed_instances` and `validation_warnings` to ` be `{index: => i, record: => r}` so that the ActiveRecord object is still being returned within `results.failed_instances`

Is returning the ActiveRecord object as part of the objects in `result.failed_instances` ok instead of changing the return type of the method?